common/src/test/java/org/conscrypt/javax/crypto/CipherBasicsTest.java

```diff
@@ -240,6 +243,94 @@ public void testAeadEncryption() throws Exception {
         }
     }
 
+    @Test
+    public void testAeadByteBufferEncryption() throws Exception {
+        TestUtils.assumeAEADAvailable();
+        for (Provider p : Security.getProviders()) {
+            for (Map.Entry<String, String> entry : AEAD_CIPHER_TO_TEST_DATA.entrySet()) {
+                String transformation = entry.getKey();
+
+                Cipher cipher;
+                try {
+                    cipher = Cipher.getInstance(transformation, p);
+                } catch (NoSuchAlgorithmException e) {
+                    // This provider doesn't provide this algorithm, ignore it
+                    continue;
+                }
+
+                List<String[]> data = readCsvResource(entry.getValue());
+                for (String[] line : data) {
+                    Key key = new SecretKeySpec(toBytes(line[KEY_INDEX]),
+                            getBaseAlgorithm(transformation));
+                    byte[] iv = toBytes(line[IV_INDEX]);
+                    byte[] _plaintext = toBytes(line[PLAINTEXT_INDEX]);
+                    byte[] _ciphertext = toBytes(line[CIPHERTEXT_INDEX]);
+                    byte[] tag = toBytes(line[TAG_INDEX]);
+                    byte[] aad = toBytes(line[AAD_INDEX]);
+
+                    ByteBuffer plaintext = ByteBuffer.wrap(_plaintext);
+
+                    // Some ChaCha20 tests include truncated tags, which the Java API doesn't
+                    // support.  Skip those tests.
+                    if (transformation.startsWith("ChaCha20") && tag.length < 16) {
+                        continue;
+                    }
+
+                    AlgorithmParameterSpec params;
+                    if (transformation.contains("GCM")) {
+                        params = new GCMParameterSpec(8 * tag.length, iv);
+                    } else {
+                        params = new IvParameterSpec(iv);
+                    }
+
+                    try {
+                        cipher.init(Cipher.ENCRYPT_MODE, key, params);
+                        if (aad.length > 0) {
+                            cipher.updateAAD(aad);
+                        }
+                        byte[] _combinedOutput = new byte[_ciphertext.length + tag.length];
+
+                        assertEquals("Provider " + p.getName()
+                                        + ", algorithm " + transformation
+                                        + " reported the wrong output size",
+                                _combinedOutput.length, cipher.getOutputSize(_plaintext.length));
+                        System.arraycopy(_ciphertext, 0, _combinedOutput, 0, _ciphertext.length);
+                        System.arraycopy(tag, 0, _combinedOutput, _ciphertext.length, tag.length);
+
+                        ByteBuffer combinedOutput = ByteBuffer.wrap(_combinedOutput);
```

Comment:
How about factoring out from ~ here to the end of the try block into a separate method, then do the same for `testAeadEncryption()` and hopfully you should be able to merge the two methods and avoid a lot of duplication.

Reply:
I factored out the separate try blocks into two methods so then the testAeadEncryption() just calls both in its internal try block